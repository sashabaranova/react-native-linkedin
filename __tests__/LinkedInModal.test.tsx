import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import LinkedInModal, {
  cleanUrlString,
  getCodeAndStateFromUrl,
  getErrorFromUrl,
  transformError,
  isErrorUrl,
  getAuthorizationUrl,
  getPayloadForToken,
  fetchToken,
  logError,
  onLoadStart,
} from '../src'

jest.mock('react-native-webview/lib/WebView', () => 'WebView')

// @ts-ignore
global.fetch = jest.fn().mockImplementation(
  () =>
    new Promise(resolve => {
      resolve({
        ok: true,
        json: () => ({
          access_token: 'access_token',
          expires_in: 'expires_in',
        }),
      })
    }),
)

it('<LinkedInModal /> render correctly', () => {
  const tree = renderer
    .create(
      <LinkedInModal
        onSuccess={() => {}}
        clientID="clientID"
        clientSecret="clientSecret"
        redirectUri="https://xaviercarpentier.fr"
        authState="authState"
        areaTouchText={{bottom: 10, left: 10, right: 10, top: 10}}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('cleanUrlString', () => {
  expect(cleanUrlString('https://xaviercarpentier.com#!')).toBe(
    'https://xaviercarpentier.com',
  )
  expect(cleanUrlString('https://xaviercarpentier.com')).toBe(
    'https://xaviercarpentier.com',
  )
})

it('getCodeAndStateFromUrl', () => {
  expect(
    getCodeAndStateFromUrl('https://xaviercarpentier.com?code=code&state=1234'),
  ).toMatchObject({ code: 'code', state: '1234' })
})

it('isErrorUrl', () => {
  expect(
    isErrorUrl(
      'https://xaviercarpentier.com?error=error&error_description=error_description',
    ),
  ).toBe(true)
})

it('getErrorFromUrl', () => {
  expect(
    getErrorFromUrl(
      'https://xaviercarpentier.com?error=error&error_description=error_description',
    ),
  ).toMatchObject({ error: 'error', error_description: 'error_description' })
})

it('transformError', () => {
  expect(
    transformError({
      error: 'error',
      error_description: 'error_description',
    }),
  ).toMatchObject({
    type: 'error',
    message: 'error_description',
  })
})

it('getAuthorizationUrl', () => {
  expect(
    getAuthorizationUrl({
      authState: 'authState',
      clientID: 'clientID',
      permissions: ['r_basicprofile', 'r_emailaddress'],
      redirectUri: 'https://xaviercarpentier.com',
    }),
  ).toBe(
    'https://www.linkedin.com/oauth/v2/authorization?' +
      'client_id=clientID&redirect_uri=https%3A%2F%2Fxaviercarpentier.com&' +
      'response_type=code&scope=r_basicprofile%20r_emailaddress&' +
      'state=authState',
  )
})

it('getPayloadForToken', () => {
  expect(
    getPayloadForToken({
      clientID: 'clientID',
      clientSecret: 'clientSecret',
      code: 'code',
      redirectUri: 'https://xaviercarpentier.com',
    }),
  ).toBe(
    'client_id=clientID&client_secret=clientSecret&' +
      'code=code&grant_type=authorization_code&' +
      'redirect_uri=https%3A%2F%2Fxaviercarpentier.com',
  )
})

it('fetchToken', async () => {
  const token = await fetchToken('payload')
  expect(token).toMatchObject({
    access_token: 'access_token',
    expires_in: 'expires_in',
  })
})

it('logError', async () => {
  logError({ type: 'test_error', message: 'test error' })
})

it('onLoadStart error', async () => {
  await onLoadStart(
    'http://url.com?error=error',
    '',
    () => {},
    (error: any) => expect(error).toEqual({ type: 'error', message: '' }),
    () => {},
    () => new Promise(resolve => resolve({})),
  )
})

it('onLoadStart success', async () => {
  await onLoadStart(
    'http://url.com?access_token=access_token&expires_in=123',
    '',
    success =>
      expect(success).toEqual({
        access_token: 'access_token',
        expires_in: '123',
      }),
    () => {},
    () => {},
    () => new Promise(resolve => resolve({})),
    true,
  )
})

it('onLoadStart error code & state', async () => {
  await onLoadStart(
    'http://url.com?access_token=access_token&expires_in=123&state=123',
    '456',
    () => {},
    error =>
      expect(error).toEqual({
        type: 'state_not_match',
        message: 'state is not the same 123',
      }),
    () => {},
    () => new Promise(resolve => resolve({})),
  )
})
