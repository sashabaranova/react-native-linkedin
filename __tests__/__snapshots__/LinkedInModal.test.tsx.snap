// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkedInModal /> render correctly 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 10,
        "left": 10,
        "right": 10,
        "top": 10,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Login with LinkedIn
    </Text>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.5)",
            "flex": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 40,
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderColor": "rgba(0, 0, 0, 0.6)",
              "borderRadius": 5,
              "borderWidth": 10,
              "flex": 1,
            },
            Object {},
          ]
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000",
            "borderColor": "rgba(255, 255, 255, 0.5)",
            "borderRadius": 12,
            "borderWidth": 2,
            "height": 24,
            "justifyContent": "center",
            "opacity": 1,
            "position": "absolute",
            "right": 5,
            "top": 35,
            "width": 24,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../src/assets/x-white.png",
            }
          }
          style={
            Object {
              "height": 16,
              "width": 16,
            }
          }
        />
      </View>
    </View>
  </Modal>
</View>
`;
